src/main/java/com/rtb/projectmanagementtool/task/TaskController.java

```diff
@@ -41,45 +53,64 @@ public TaskData getTaskByID(long taskID) {
     return task;
   }
 
-  public ArrayList<TaskData> getTasks(int quantity, String sortBy, String sortDirection) {
-    ArrayList<TaskData> tasks = new ArrayList<>();
-    Query query;
-    if (sortDirection.equals("descending")) {
-      query = new Query("Task").addSort(sortBy, SortDirection.DESCENDING);
-    } else {
-      query = new Query("Task").addSort(sortBy, SortDirection.ASCENDING);
-    }
-    return getTasks(query, quantity);
+  public ArrayList<TaskData> getTasks(int limit, String sortBy, String sortDirection) {
+    SortPredicate sort =
+        new SortPredicate(sortBy, SortDirection.valueOf(sortDirection.toUpperCase()));
+    return getTasks(NO_QUERY_FILTER, limit, sort);
   }
 
   public ArrayList<TaskData> getSubtasks(TaskData task) {
-    ArrayList<Key> subtaskKeys = getKeysFromTaskIDs(task.getSubtasks());
-    if (!task.getSubtasks().isEmpty()) {
-      Query query = new Query("Task").addFilter("__key__", FilterOperator.IN, subtaskKeys);
-      return getTasks(query, Integer.MAX_VALUE);
-    }
-    return new ArrayList<>();
+    Filter filter = new FilterPredicate("parentTaskID", FilterOperator.EQUAL, task.getTaskID());
+    return getTasks(filter, NO_QUERY_LIMIT, NO_QUERY_SORT);
   }
 
-  private ArrayList<TaskData> getTasks(Query query, int quantity) {
+  private ArrayList<TaskData> getTasks(Filter filter, int limit, SortPredicate sort) {
     ArrayList<TaskData> tasks = new ArrayList<>();
+    Query query = new Query("Task");
+    if (filter != null) {
+      query.setFilter(filter);
+    }
+    if (sort != null) {
+      query.addSort(sort.getPropertyName(), sort.getDirection());
+    }
     PreparedQuery results = datastore.prepare(query);
     int count = 0;
     for (Entity entity : results.asIterable()) {
-      if (count++ >= quantity) {
+      if (count >= limit) {
         break;
       }
-      TaskData task = new TaskData(entity);
-      tasks.add(task);
+      tasks.add(new TaskData(entity));
+      count++;
     }
     return tasks;
   }
 
+  // Delete methods
+
   public void deleteTasks(ArrayList<Long> taskIDs) {
-    ArrayList<Key> keys = getKeysFromTaskIDs(taskIDs);
-    datastore.delete(keys);
+    TransactionOptions options = TransactionOptions.Builder.withXG(true);
+    Transaction transaction = datastore.beginTransaction(options);
+    try {
+      deleteTasksRecursive(taskIDs);
+      transaction.commit();
+    } finally {
```

Comment:
Should this be `catch` instead of `finally`?   iiuc finally is executed every time the try block exits.  Which means on successful operations, this would also rollback?

Reply:
In the documentation it used `finally`. Maybe the transaction isn't considered active after committing? It passes the unit tests though.

Reply:
It could be possible it would auto close right after the commit.  In which case the finally would be ok.